eth/stagedsync/stage_headers.go

```diff
@@ -135,12 +124,26 @@ func HeadersPOS(
 	cfg HeadersCfg,
 	initialCycle bool,
 	test bool, // Set to true in tests, allows the stage to fail rather than wait indefinitely
-	useExternalTx bool,
 ) error {
+	// We need to have another write transaction for the miner in case we want to validate.
+	if tx != nil {
+		if err := tx.Commit(); err != nil {
+			return err
+		}
+	}
 	atomic.StoreUint32(cfg.waitingPosHeaders, 1)
 	// Waiting for the beacon chain
 	log.Info("Waiting for payloads...")
-	header := <-cfg.reverseDownloadCh
+	payloadMessage := <-cfg.reverseDownloadCh
```

Comment:
I'd move the waitingPosHeaders manipulations closer together to reduce the chance of an incorrect value. I.e. something like
```
...
log.Info("Waiting for payloads...")
atomic.StoreUint32(cfg.waitingPosHeaders, 1)
payloadMessage := <-cfg.reverseDownloadCh
atomic.StoreUint32(cfg.waitingPosHeaders, 0)
header := payloadMessage.Header
...
```